### ISSUE TITLE
get_sell_rate() using bids instead of asks?

### ISSUE THREAD

**Issue by yazeed** [freqtrade/freqtrade#2907]
Hello,

So, I just noticed that get_sell_rate() method uses bids instead of asks, and I wasn't sure why. Logging this to discuss this as I'm not sure why it's implemented this way. Talking to @hroff-1902 he mentioned that he'd eventually like to see both get_buy_rate and get_sell_rate refactored to be used differently with limit/market orders, so perhaps this is an opportunity to discuss this.

There were also several discussions somewhat related to this made in #1667 and #2637, but I wasn't sure they were entirely focused on using bids in get_sell_rate.

```
def get_sell_rate(self, pair: str, refresh: bool) -> float:
        """
        Get sell rate - either using get-ticker bid or first bid based on orderbook
        The orderbook portion is only used for rpc messaging, which would otherwise fail
        for BitMex (has no bid/ask in fetch_ticker)
        or remain static in any other case since it's not updating.
        :return: Bid rate
        """
        config_ask_strategy = self.config.get('ask_strategy', {})
        if config_ask_strategy.get('use_order_book', False):
            logger.debug('Using order book to get sell rate')
            order_book = self.exchange.get_order_book(pair, 1)
            rate = order_book['bids'][0][0]
        else:
            rate = self.exchange.fetch_ticker(pair, refresh)['bid']
        return rate
```

**Comment by hroff-1902** [comment]
> Talking to @hroff-1902 he mentioned that he'd eventually like to see both get_buy_rate and get_sell_rate refactored to be used differently with limit/market orders.

Well, these were some thoughts, not particular proposals. 😉 
However, right, as I said, I do have some serious doubts regarding current behavior, so this issue is a place to discuss it.

**Comment by yazeed** [comment]
Would love to hear more about this, as I am trying to understand different fundamental pieces of freqtrade's code.

**Comment by xmatthias** [comment]
#1059 seems to hint at something similar - but it's clear from the issue - as the proposed solution includes "putting orders early and leaving them hanging".

However, for a recent explanation, best have a look through #2792.
While the title suggests something else, it's mostly about buy rates ... and why / how they are calculated.

It's not a bug for sure as it does work as intended / designed. 

There are arguments for both sides - using either asks or bids for pricing buys.
the current method will lower your profit by spread  -the alternative is a lower rate of being filled.

But by using bids when buying - and ask price when selling, you'll significantly lower the 
probability of being filled on a limit order - to the most extreme case if you've got a very good strategy which did predict a breakout NOW - it would miss on that breakout - which in the end must be seen as lost profit as the strategy was right - it just wasn't filled and price "never looked back".

**Comment by yazeed** [comment]
> #1059 seems to hint at something similar - but it's not clear from the issue - as the proposed solution includes "putting orders early and leaving them hanging".
> 
> However, for a recent explanation, best have a look through #2792.
> While the title suggests something else, it's mostly about buy rates ... and why / how they are calculated.
> 
> It's not a bug for sure as it does work as intended / designed.
> 
> There are arguments for both sides - using either asks or bids for pricing buys.
> the current method will lower your profit by spread -the alternative is a lower rate of being filled.
> 
> But by using bids when buying - and ask price when selling, you'll significantly lower the
> probability of being filled on a limit order - to the most extreme case if you've got a very good strategy which did predict a breakout NOW - it would miss on that breakout - which in the end must be seen as lost profit as the strategy was right - it just wasn't filled and price "never looked back".

So your rationale is because it has better chances at entry/exits, even if that was at the cost of the spread? Why not just user market orders, then? To me, this still sounds like a bug as get_sell_rate should look at the top asks to be as intended.

**Comment by xmatthias** [comment]
> There are other means for getting orders filled faster such as orderbook and flow monitoring.

Well yes - but also when you look at the orderbook, you need to decide at which side of the orderbook you look at when buying or selling - using orderbook or `fetch_ticker()` won't differ very much - as both will (should) return the best ask/bid price at that given time.

> To me, this still sounds like a bug as get_sell_rate() should look at the top asks to be as intended, to be true to its own definition as a function. 

Why? `get_sell_rate()` returns the rate at which to sell the trade. Nowhere in the definition does it say at which side of the orderbook it looks for that rate - so there is no "intended" here.

Also - who says that it's correct to look at the best ask when defining a sell price? 

> Freqtrade currently returns buy rates for both get_sell_rate() and get_buy_rate().

This is definitely not true. Assuming default configurations (orderbook is off).

`get_sell_rate()` returns the `bid` value of `fetch_tickers()`
`get_buy_rate()` returns the `ask` side, or something between `ask` and `last` (depending on the configuration) - again - from `fetch_tickers()`. (look at https://github.com/freqtrade/freqtrade/issues/2792#issuecomment-577187715 for a detailed explanation of that).


So if any - then both functions are wrong / inversed.

----

If you read the linked comment carefully (and to the end) - you will note that i am NOT certain that the side of the orderbook we're looking at is the correct one - but i'm also not certain that it's the wrong one.

And that's the real problem to be me - there is no definitive "correct" or wrong here.
Assume a manual trading process - you may look at `asks` for selling - while i do look at `bids` for selling (or viceversa) - that's a different way to price your trades - but which approach is proven to be wrong? I've not yet seen such a definition written ...

**Comment by xmatthias** [comment]
I guess the best option is to make this easily configurable for the user (since opinions on this will differ ...).

I've got a PR almost ready (right after #2954 is merged as it's basedon top of that branch but would like to keep them seperate).

--- 

I think the key problem is the inconsistency we got at the moment - looking at `get_buy_rate()` for example, we use the **bid side** if orderbook is enabled, but we use the **ask side** if orderbook is disabled.

the same is true for `get_sell_rate()` - however the problem here is slightly different as the orderbook section for this is only used for "reporting" - but not to generate the exit price (that part is in `handle_trade()` - as it's using a loop testing multiple orderbook positions).
